### ISSUE TITLE
Tools to align objects

### ISSUE THREAD

**Issue by d1g** [openstreetmap/iD#2987]
Probably JOSM tools are too low level (basic align nodes in line and such).

Inkscape gets it okay https://inkscape.org/en/gallery/item/2504/
but Photoshop and CAD tools get it even better by automatically highlighting anchor nodes when _simply moving object_ in editor (I'm not able to find good illustration ATM)

We also need toggle button to speed up editor and more precise positioning (when we need it). You may create panel/multi-button "precise tools/modes".

**Comment by bhousel** [comment]
Yeah I find myself wanting this sometimes too.  The ability to place guidelines and align objects to them would be really useful.

**Comment by g4rpch** [comment]
really useful tool please add it on ID!!

Thanks a lot
G4rp

**Comment by homersimpsons** [comment]
As a workaround, you can use lines:
- create a ligne that is connected to all the nodes you want to align
- place the first one and the last one node where there should be 
- click 's' (straight line) 
- delete the line

It's not really much longer. But it's easy and this avoid to implement new functionality.

**Comment by g4rpch** [comment]
nice workaround!
But if you can implement the functionality also users not expert can use it ;)

**Comment by jidanni** [comment]
Tool idea: The rectangle tool is great. But after making a house's
sides align a perfect right angles, one would also often want that
house then again to align perfectly to the nearby road, and neigbors'
houses.

Sure in newer communities plots and homes might have varying
orientations. But in many traditional towns one would like to make the
first house we draw align to the road in front. (Keep the same
distance, but swing the house from say 89 degree angle (the best we
can do with our mouse hand and eyes) to exact 90 degrees.)

Then the next house the same, perhaps not only aligning with the
nearest road, but also with the nearest house...

Anyway perhaps call this tool "align with nearby linear features"
tool, and have it appear only after one has squared a house... or have
it always available for use with houses and roads themselves...
()There probably is already a way to first lay down the southbound lanes
of an expressway. Then add the parallel but somewhat distant
northbound lanes.)

**Comment by mmd-osm** [comment]
I think this is already possible with iD 3, iirc, @quincylvania demoed something similar at SotM US 2019. You should watch the video!

**Comment by quincylvania** [comment]
> I think this is already possible with iD 3, iirc, @quincylvania demoed something similar at SotM US 2019. You should watch the video!

Hmm I don't think so! I demoed a preliminary drawing tool that enforces square corners, but not alignment like this 😄

**Comment by jidanni** [comment]
(E.g., https://www.openstreetmap.org/changeset/78554138
After I drew and squared the building, I drew and squared the
parking lots. So far so good. But then I had to rotate the parking
lots with only my eye to help them from not ending up (too) crooked,
compared to each other, and the building (which itself I hope isn't
too crooked relative the road. (One would expect true parallel lines
here in city-planned areas.)))

**Comment by jidanni** [comment]
![25545-1](https://user-images.githubusercontent.com/1230959/71137783-3923f600-2244-11ea-8196-1a5f07aad849.jpg)
I am saying "help me ensure all the lines in the picture here have
zero rotation compared with each other. (And all based on the road, which was there first
before we started drawing.)
What about the north parking lot being pulled a little northwest of
the south parking lot? Well that could be an additional bonus command,
(Make my stack of bricks more match the neighboring kid in the
playlot's stack of bricks. And exactly in what ways?)
But at least first work on a rotation lock...

**Comment by jidanni** [comment]
And here it would be great if there was a tool to help me make the south lanes bridge run parallel to the north lanes bridge, instead of trying to do one's best with the mouse.
![27315-6](https://user-images.githubusercontent.com/1230959/71146175-20750980-225f-11ea-9308-cf1f483a1696.jpg)

**Comment by jidanni** [comment]
I've got it! It's so simple.
```
               C
      B        Y      D
      X               Z

A                             E
```
1. User creates first line, with nodes A, B, C, D, E.
1. Now user starts a second line, with node X say 5 meters from B.
1. Now he hovers the mouse below D. A pulsating point appears 5 meters
from D.
1. If he clicks it, it becomes Z, and new parallel line XYZ is
automatically finished!
1. In fact back at the pulsating step you could have the whole new
potential line and nodes pulsate.

Let's say he really just wanted a straight line XZ with no Y. Well then
he needn't click exactly on the pulsating Z offered, but nearby,
upon which he would just get his plain XZ line (with his placement of Z,
not the one we offered him, with no Y being created.)
